tools/meteor-services/deploy.js

```diff
@@ -321,6 +334,122 @@ function canonicalizeSite(site) {
+  // Create a default polling configuration for polling for deploy / build
+  // In the future, we may change this to be user-configurable or smart
+  // The user can only currently configure the polling timeout via a flag
+  let pollingState = new PollingState(deployPollTimeout);
+  
+  var deploymentPollResult = buildmessage.enterJob({
+    title: "Waiting for deploy to succeed..."},
+    function() {
+      return pollForDeploy(pollingState,
+        versionId);
+    });
+
+  if (deploymentPollResult && deploymentPollResult.isActive) {
+    return 0;
+  }
+  return 1;
+}
+
+// Creates a polling configuration with defaults if fields left unset
+// Right now we only use the default unless timeout is specified
+// We envision potentially creating this configuration object in a programmatic
+// way or via user-specification in the future.
+// Default initialWaitTime is 10 seconds – this is the time to wait before checking at all
+// Default pollInterval is 700 milliseconds – this is the wait interval between polls
+// Default timeout is 15 minutes
+class PollingState {
+  constructor(timeoutMs,
+    initialWaitTimeMs,
+    pollIntervalMs) {
+      const fifteenMinutesMs = 15*60*1000;
+      this.initialWaitTimeMs = initialWaitTimeMs || 10*1000;
+      this.pollIntervalMs = pollIntervalMs || 700;
+      this.timeoutMs = timeoutMs || fifteenMinutesMs
+  }
+
+  // Declare tracking variables for the duration of the poll
+  // `start` tracks the time when we started polling
+  // `currentMessage` tracks what the current status message is for this version
+  initialize() {
+    this.start = new Date();
+    this.currentMessage = '';
+  }
+}
+
+// Poll the "version-status" endpoint for the build and deploy status
+// of a specified version ID with a polling configuration.
+// This will only end successfully when the polling endpoint reports that
+// the version deployment is finished. The version-status endpoints will report
+// messages pertaining to the status of the version, which will then be reported
+// directly to the user. When the poll is complete, it will return an object
+// with information about the final state of the version and the app.
+function pollForDeploy(pollingState, versionId, firstRun = true) {
+  const {
+    timeoutMs,
+    initialWaitTimeMs,
+    pollIntervalMs,
+    start,
+    currentMessage,
+  } = pollingState;
+  // If this is the first poll, we will wait for initialWaitTimeMs milliseconds 
+  // before beginning. We also want to initialize all of our tracking variables
+  if(firstRun) {
+    // Initialize the state of the poll, then begin polling after a pause
+    pollingState.initialize();
+    sleepMs(initialWaitTimeMs);
+    return pollForDeploy(pollingState, versionId, false);
+  }
+
+  // Do a call to the version-status endpoint for the specified versionId
+  const versionStatusResult = deployRpc({
+    method: 'GET',
+    operation: 'version-status',
+    site: versionId,
+    expectPayload: [],
```

Comment:
You can list 'message' and 'finishStatus' here and then you don't need to do as much error handling later

Reply:
👍